### PR TITLE
ENH Layer/model status shows devices now

### PR DESCRIPTION
**Pull request by BenjaminBossan**
For each adapter, show all the devices of this adapter's parameters.

Also, while working on this, found a very minor bug in VeRA as its linear layer didn't implement its own `__repr__`.

### CODE REVIEW

File: src/peft/peft_model.py [hunk 1]
```diff
@@ -2510,6 +2513,20 @@ def check_irrgular(vals: list[bool]) -> bool | Literal["irregular"]:
...
+        devices_dd = collections.defaultdict(list)
+        for adapter_module_name in module.adapter_layer_names + module.other_param_names:
+            adapter_module = getattr(module, adapter_module_name)
+            if isinstance(adapter_module, torch.nn.ModuleDict):
+                for key, submodule in adapter_module.items():
+                    devices_dd[key].extend([param.device.type for param in submodule.parameters()])
+            elif (
+                isinstance(adapter_module, torch.nn.ParameterDict)
+                or (adapter_module.__class__.__name__ == "BufferDict")  # VeRA
+            ):
```
Comment (sayakpaul) [inline src/peft/peft_model.py:2525]: Should this be covered for DoRA too?

**Review by sayakpaul — APPROVED** [review]
Just a single comment. I like this feature a lot!

The tests are very comprehensive. 

Would you mind adding a note about this (once this is merged) to https://huggingface.co/docs/diffusers/main/en/tutorials/using_peft_for_inference?

File: docs/source/developer_guides/troubleshooting.md [hunk 2]
```diff
@@ -214,21 +214,23 @@ It is possible to get this information for non-PEFT models if they are using PEF
...
+                  available_adapters=['adapter-1', 'adapter-2'],
+                  devices={'adapter-1': ['cpu'], 'adapter-2': ['cuda]}),
  TunerLayerStatus(name='text_model.encoder.layers.0.self_attn.v_proj',
                   module_type='lora.Linear',
                   enabled=True,
                   active_adapters=['adapter-2'],
                   merged_adapters=[],
                   requires_grad={'adapter-1': False, 'adapter-2': True},
-                  available_adapters=['adapter-1', 'adapter-2']),
+                  devices={'adapter-1': ['cpu'], 'adapter-2': ['cuda]}),
```
Comment (pacman100) [inline docs/source/developer_guides/troubleshooting.md:233]:
```suggestion
                  devices={'adapter-1': ['cpu'], 'adapter-2': ['cuda']},
```

File: docs/source/developer_guides/troubleshooting.md [hunk 3]
```diff
@@ -244,5 +246,6 @@ TunerModelStatus(
     merged_adapters=[],
     requires_grad={'adapter-1': False, 'adapter-2': True},
     available_adapters=['adapter-1', 'adapter-2'],
+    devices={'adapter-1': ['cpu'], 'adapter-2': ['cuda],
```
Comment (pacman100) [inline docs/source/developer_guides/troubleshooting.md:249]:
```suggestion
    devices={'adapter-1': ['cpu'], 'adapter-2': ['cuda']},
```

**Review by pacman100 — APPROVED** [review]
Thank you @BenjaminBossan for adding the features for showing model and layer adapter details and improving that to add the device details! Left a couple nits.

File: src/peft/peft_model.py (hunk 1 above)
Comment (BenjaminBossan) [reply src/peft/peft_model.py:2525]: Yes, this works for DoRA too. As DoRA is just a `LoraModel`, the shown status will be almost the same for LoRA and DoRA, except that DoRA has a few more parameters.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot !

**Pull request merged**